indico_toolkit/indico_wrapper/dataset.py

```diff
@@ -68,42 +69,73 @@ def create_empty_dataset(
-            verbose (bool, optional): print updates as each batch processes. Defaults to True.
+            num_threads (int, optional): number of threads to use for file upload. Defaults to 3.
+            upload_batch_size (int, optional): number of files to upload at a time. Defaults to 3.
+            process_batch_size (int, optional): number of files to process at a time. Defaults to 5.
+            max_download_checks (int, optional): number of times to poll for uploads to complete.
+                                                 Necessary because documents can get stuck in a DOWNLOADING
+                                                 state. Defaults to 10 (~1 minute)
         """
         fp = FileProcessing(filepaths)
         dataset_id = self.create_empty_dataset(dataset_name).id
-        number_of_batches = int(math.ceil(len(filepaths) / batch_size))
-        for batch_fpaths in fp.batch_files(batch_size):
-            dataset = self.add_files_to_dataset(dataset_id, batch_fpaths)
-            if verbose:
-                number_of_batches -= 1
-                print(
-                    f"Finished batch at {datetime.now().time()}: {number_of_batches} remaining"
+
+        # Upload using an executor - take advantage of multiple uploaders on the platform
+        results = self._upload_threaded(fp, num_threads, upload_batch_size)
+        print("Uploaded Doucments")
+
+        add_files_batch_size = 100
+        for i in tqdm(range(0, len(results), add_files_batch_size)):
+            self.client.call(
+                _AddFiles(
+                    dataset_id=dataset_id,
+                    metadata=results[i : i + add_files_batch_size],
                 )
-        return dataset
+            )
+        print("Added Files to dataset, waiting for documents to download")
+        dataset = self.client.call(GetDatasetFileStatus(id=dataset_id))
 
-    def create_dataset(self, filepaths: List[str], dataset_name: str) -> Dataset:
-        dataset = self.client.call(
-            CreateDataset(
-                name=dataset_name,
-                files=filepaths,
+        debouncer = Debouncer(max_timeout=max_download_checks)
+        while self._datafiles_downloading(dataset, debouncer, max_download_checks):
+            dataset = self.client.call(GetDatasetFileStatus(id=dataset_id))
+            debouncer.backoff()
+
+        print("Downloads complete, starting pdf extraction jobs")
+        file_ids = [df.id for df in dataset.files if df.status == "DOWNLOADED"]
+        for i in tqdm(range(0, len(file_ids), process_batch_size)):
+            self.client.call(
+                _ProcessFiles(
+                    dataset_id=dataset_id,
+                    datafile_ids=file_ids[i : i + process_batch_size],
+                )
             )
-        )
-        self.dataset_id = dataset.id
+        print("Files processing")
+        dataset = self.client.call(GetDatasetFileStatus(id=dataset_id))
```

Comment:
shouldn't this just be GetDataset?

Reply:
GetDataset from the SDK does not return the statuses of the uploaded files

Reply:
ahh- so maybe var should be dataset_statuses or something, on quick read thought you were just returning hte dataset object